spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/security/oauth2/client/ClientRegistrationRepositoryAutoConfiguration.java

```diff
@@ -0,0 +1,146 @@
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationProperties;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository;
+import org.springframework.util.Assert;
+import org.springframework.util.ObjectUtils;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for a <code>ClientRegistrationRepository</code> bean,
+ * which serves as a store for OAuth 2.0 / OpenID Connect 1.0 client registrations.
+ *
+ * @author Joe Grandja
+ * @since 2.0.0
+ * @see ClientRegistrationRepository
+ */
+@Configuration
+@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
+@ConditionalOnClass({ EnableWebSecurity.class, ClientRegistration.class })
+@EnableConfigurationProperties(OAuth2Properties.class)
+public class ClientRegistrationRepositoryAutoConfiguration {
+
+	@Configuration
+	@Conditional(ClientsConfiguredCondition.class)
+	@ConditionalOnMissingBean(ClientRegistrationRepository.class)
+	protected static class ClientRegistrationRepositoryConfiguration {
+
+		@Bean
+		public ClientRegistrationRepository clientRegistrationRepository(
+				Environment environment, OAuth2Properties oauth2Properties) {
+
+			this.mergeClientDefaults(environment, oauth2Properties);
+
+			List<ClientRegistration> clientRegistrations = new ArrayList<>();
+
+
+			oauth2Properties.getClients().values().stream()
+					.filter(e -> !ObjectUtils.isEmpty(e.getClientId()))
+					.collect(Collectors.toList())
+					.forEach(clientProperties -> clientRegistrations.add(
+							new ClientRegistration.Builder(clientProperties).build()));
+
+			return new InMemoryClientRegistrationRepository(clientRegistrations);
+		}
+
+		private void mergeClientDefaults(Environment environment, OAuth2Properties oauth2Properties) {
+			Assert.isInstanceOf(ConfigurableEnvironment.class, environment, "environment must be a ConfigurableEnvironment");
+
+			MutablePropertySources propertySources = new MutablePropertySources();
+			((ConfigurableEnvironment) environment).getPropertySources().forEach(propertySource -> {
+				if (MapPropertySource.class.isAssignableFrom(propertySource.getClass())) {
+					propertySources.addLast(propertySource);
+				}
+			});
+			propertySources.addLast(ClientPropertiesUtil.loadDefaultsPropertySource());
+
+			Map<String, ClientRegistrationProperties> mergedClients =
+					new Binder(ConfigurationPropertySources.from(propertySources))
+							.bind(ClientPropertiesUtil.CLIENTS_PROPERTY_PREFIX,
+									Bindable.mapOf(String.class, ClientRegistrationProperties.class))
+							.get();
+
+			// Override with merged client properties
+			oauth2Properties.setClients(mergedClients);
```

Comment:
IMO this is too low-level. 

What if we had a `Map<String,Foo>` (name TBD) and one attribute of this object would be a the ID of an oauth2 client type (`google`, `github`, `facebook`, etc). What the auto-configuration can then do is apply default values based on the client-type. If the client is fully configured, that's fine and we'll use that but all the properties that can be inferred by the client type are then set.

In Spring Boot we could put something in place so that new "id" can be registered if necessary.

Reply:
I'm not sure if I completely follow? Maybe an example to elaborate would help.

Reply:
I think what @snicoll is suggesting is an attribute that specifies the "template" for common properties.

We could have an enum (probably in Spring Security) that defines well known templates:

```java
public enum OAuth2ClientTemplate {

    GOOGLE, FACEBOOK, GITHUB

}
```

In the client property we then have a "template" field:

```
public class OAuth2ClientProperties {
    private OAuth2ClientTemplate template;
    private String redirectUrl;
    private String scope;
    // etc
}
```

The user can then specify the template in the properties file. Some examples:

```properties
# Standard GitHub with some keys
spring.security.oauth2.client.mygithub.template=GITHUB
spring.security.oauth2.client.mygithub.token=ZYX

# GitHub but override the scope with some keys
spring.security.oauth2.client.foogithub.template=GITHUB
spring.security.oauth2.client.foogithub.token=ZYX
spring.security.oauth2.client.foogithub.scope=email

# Something completely custom
spring.security.oauth2.client.bar.redirect-uri=...
spring.security.oauth2.client.bar.token=ZYX
spring.security.oauth2.client.barscope=baz
```

Reply:
I see. Let me give this one some thought

Reply:
That's what I meant (with a few things we may need to adapt but we shouldn't worry about that now). The main idea is that this mechanism of providing default values should be in Spring Session and therefore the API that allows you to configure your client should be there as well. 

That's also a reason why we shouldn't bind Spring Session's API directly on the environment. We need to be in control on what options we expose to the user.